### ISSUE TITLE
Charon Server?

### ISSUE THREAD

**Issue by KeyPuncher** [gamedevware/charon-unity3d#8]
So this is just the Unity plugin, no code for the local server?
I was hoping to add inheritance.

**Comment by deniszykov** [comment]
Hi @KeyPuncher !
Yes, server part is closed source.
If you mean the inheritance of entities. Then it is could be implemented using composition.
For example Item->Armor, Item -> Armor inheritance via composition.
```
Item
 * Id
 * Name

Armor
 * Item
 * Amount

Weapon
 * Item
 * Damage
```
For ease of use, additional methods and fields can be generated for these ''inherited' classes in partial code:
```cs
	partial class Item
	{
                public string ItemType { get; set; }
        }

	partial class GameData
	{
		partial void OnInitialize()
		{
		    	foreach (var armor in this.allArmor)
			{
                            	armor.Item.ItemType = "Armor"
                        }
		    	foreach (var weapon in this.allWeapon)
			{
                            	weapon.Item.ItemType = "Weapon"
                        }
		}
        }

```

# Advanced stuff
Also you could modify T4 tempates for code generation, and produce fully customized code.
[Export T4 Templates](https://github.com/deniszykov/charon-unity3d/wiki/Exporting%20Code%20Generation%20Templates)

Then edit GameDataExample.tt file to provide right path to gamedata file.

## IDE Way
Use VS (MonoDevelop) to process (<right click on GameDataExample.tt file> -> Run Custom Tools) templates into source code.

## Deploy Script Way

Run Code Generation (not documented in wiki)
```sh
Charon.exe GENERATE TEXT --templatePath GameDataExample.tt --outputPath "c:/my/gamedata.cs"

# show help for GENERATE TEXT command
Charon.exe GENERATE HELP TEXT
```

**Comment by KeyPuncher** [comment]
I see, with composition I'd have to go up the hierarchy to link everything and validate the gamedata post creation. 

In my case I have a number of different Entity definitions but they are all used in the same way upon another parenting entity. Was hoping to make the linking process slightly easier than loosely linking them by Id. Using an entity like a type of linker like you suggested Item would be could work. Was just wanting to avoid the extra class and work involved entering the game data.

Thanks for the options, I may try to learn templates.
Pretty sweet tool.

**Comment by KeyPuncher** [comment]
I got it to work using templates and processing them in Visual Studio. Once I realized the .t4 files were just addons and were not meant to be processed on their own, I surfaced the Group property for the Entities/Documents and made new partial interfaces with the Group names. Though I suppose I could've also worked with just surfacing the Group name as a property of the generated partial class (which I also just realized were partial classes, so nice). 
Though I still had to delete some items within the generated file otherwise it would not compile, but it looks like those lines weren't in the CS file Charon.exe outputted so didn't look like they were needed at all.

What did not work was using the Charon.exe GENERATE command. It was behaving like it could not find the referenced import DLLs. I tried to move the correct ones directly in the path of Charon.exe but still wound up with System.Object couldn't be found or something like that.
